library/pkcs7.c

```diff
@@ -0,0 +1,708 @@
+{
+    unsigned char *start;
+    unsigned char *end;
+    size_t len = 0;
+    int ret;
+    int isoidset = 0;
+
+    if( !pkcs7 )
+        return( MBEDTLS_ERR_PKCS7_BAD_INPUT_DATA );
+
+    /* make an internal copy of the buffer for parsing */
+    pkcs7->raw.p = start = mbedtls_calloc( 1, buflen );
+    if( pkcs7->raw.p == NULL )
+    {
+        return( MBEDTLS_ERR_PKCS7_ALLOC_FAILED );
+    }
+    memcpy( start, buf, buflen );
+    pkcs7->raw.len = buflen;
+    end = start + buflen;
+
+    ret = pkcs7_get_content_info_type( &start, end, &pkcs7->content_type_oid );
+    if( ret != 0 )
+    {
+        len = buflen;
+        goto try_data;
+    }
+
+    if( ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_ENCRYPTED_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_ENVELOPED_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_SIGNED_AND_ENVELOPED_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_DIGESTED_DATA, &pkcs7->content_type_oid )
+     || ! MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_ENCRYPTED_DATA, &pkcs7->content_type_oid ) )
+    {
+        ret =  MBEDTLS_ERR_PKCS7_FEATURE_UNAVAILABLE;
+        goto out;
+    }
+
+    if( MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_SIGNED_DATA, &pkcs7->content_type_oid ) )
+    {
+        ret = MBEDTLS_ERR_PKCS7_BAD_INPUT_DATA;
+        goto out;
+    }
+
+    isoidset = 1;
+    start = start + pkcs7->content_type_oid.len;
+
+    ret = pkcs7_get_next_content_len( &start, end, &len );
+    if( ret != 0 )
+        goto out;
+
+try_data:
+    ret = pkcs7_get_signed_data( start, len, &pkcs7->signed_data );
+    if ( ret != 0 )
+        goto out;
+
+    if ( !isoidset )
+    {
+        pkcs7->content_type_oid.tag = MBEDTLS_ASN1_OID;
+        pkcs7->content_type_oid.len = MBEDTLS_OID_SIZE( MBEDTLS_OID_PKCS7_SIGNED_DATA );
+        pkcs7->content_type_oid.p = (unsigned char *)MBEDTLS_OID_PKCS7_SIGNED_DATA;
+    }
+
+    ret = MBEDTLS_PKCS7_SIGNED_DATA;
+
+out:
+    if ( ret < 0 )
+        mbedtls_pkcs7_free( pkcs7 );
+
+    return( ret );
+}
+
+int mbedtls_pkcs7_signed_data_verify( mbedtls_pkcs7 *pkcs7,
+                                      const mbedtls_x509_crt *cert,
+                                      const unsigned char *data,
+                                      size_t datalen )
+{
+
+    int ret = MBEDTLS_ERR_PKCS7_VERIFY_FAIL;
+    unsigned char *hash;
+    mbedtls_pk_context pk_cxt = cert->pk;
+    const mbedtls_md_info_t *md_info;
+    mbedtls_md_type_t md_alg;
+    mbedtls_pkcs7_signer_info *signer;
+
+    if( pkcs7->signed_data.no_of_signers == 0 )
+        return( MBEDTLS_ERR_PKCS7_VERIFY_FAIL );
+
+
+    /*
+     * Potential TODOs
+     * Currently we iterate over all signers and return success if any of them
+     * verify.
+     *
+     * However, we could make this better by checking against the certificate's
```

Comment:
IIUC, this would mean that if the hash from a different signature (i.e., where the cert IssuerAndSerialNumber doesn't match between cert and PKCS7 signer) happened to match the given cert, we could incorrectly return success?

Reply:
I'm leaning towards saying that we should perform this check. What do other reviewers think?

Reply:
> IUC, this would mean that if the hash from a different signature (i.e., where the cert IssuerAndSerialNumber doesn't match between cert and PKCS7 signer) happened to match the given cert, we could incorrectly return success?
Yes. At the moment, the `verify` functions are not making any checks to ensure that the certificate given by the user is a match for one of the certificates in the PKCS7 `IssuerAndSerialNumber` fields. Verification is "successful" if any of the signed data items can be unsigned with the user given cert and matches the hash of the user given input. 

Daniel was also specifying in this comment that if we implement `IssuerAndSerialNumber` checks against the user given cert, we can differentiate verification failures from incorrect certs and unmatched hashes.

The only argument against this check is that any determined malicious party could easily edit the `IssuerAndSerialNumber` to inhibit the correct certificate from verifying successfully. But I will wait for more feedback before making any edits.

Reply:
@bensze01 do you have a view on this?

Reply:
On balance this is really the same problem as some random content by chance happening to have a hash such that it appears to be signed. So I am OK to skip this.